c10/core/TensorImpl.cpp

```diff
@@ -318,6 +318,12 @@ void TensorImpl::release_resources() {
   if (storage_) {
     storage_ = {};
   }
+  if (owns_pyobj_) {
+    TORCH_INTERNAL_ASSERT(pyobj_interpreter_ != nullptr);
+    TORCH_INTERNAL_ASSERT(pyobj_ != nullptr);
+    pyobj_interpreter_.load(std::memory_order_acquire)->decref(pyobj_);
+    pyobj_ = nullptr; // for safety
```

Comment:
I'm not sure, it was too hard to reason out the destruction ordering so I just slapped this one to be safe.

Reply:
After thinking more about this the ownership switch makes the ordering pretty easy actually. Because the only way for us to enter this pair of destructor is if there is no external reference to either of the objects. So nothing can reference this from other threads and chain of destructors cannot lead to these objects being used again.